### PR TITLE
CRUD on parking lots

### PR DESCRIPTION
**Pull request by ShrawanBK**
depends on: https://github.com/idmc-labs/helix-server/pull/84

### CODE REVIEW

File: src/components/Navbar/index.tsx [hunk 1]
```diff
@@ -108,6 +108,12 @@ const Navbar = (props: Props) => {
                             activeClassName={styles.active}
                             route={route.performanceAndAdmin}
                         />
+                        <SmartNavLink
+                            exact
+                            className={styles.link}
+                            activeClassName={styles.active}
+                            route={route.parkingLots}
+                        />
```
Comment (tnagorra) [inline src/components/Navbar/index.tsx:116]: Maybe move this to popupbutton

File: src/views/ParkingLots/index.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+function ParkingLots(props: ParkingLotsProps) {
+    const {
+        className,
+    } = props;
+
+    return (
+        <div className={_cs(styles.parkingLots, className)}>
+            <PageHeader
+                title="Parking Lots"
```
Comment (tnagorra) [inline src/views/ParkingLots/index.tsx:20]:
```suggestion
                title="Parking Lot"
```

File: src/views/ParkingLots/index.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import { _cs } from '@togglecorp/fujs';
+import PageHeader from '#components/PageHeader';
+import styles from './styles.css';
+import ParkingLotsTable from '#components/ParkingLotsTable';
+
+interface ParkingLotsProps {
+    className?: string;
+}
+
+function ParkingLots(props: ParkingLotsProps) {
```
Comment (tnagorra) [inline src/views/ParkingLots/index.tsx:12]:
```suggestion
function ParkingLot(props: ParkingLotProps) {
```

File: src/views/Dashboard/index.tsx [hunk 4]
```diff
@@ -35,20 +37,20 @@ function Dashboard(props: DashboardProps) {
...
+                                /* TODO: set link */
+                                <a
+                                    href="#media-monitoring-platform"
+                                >
+                                    Go to Media Monitoring Platform
+                                </a>
+                            )}
+                            defaultUser={user?.id}
+                            defaultStatus="TO_BE_REVIEWED"
+                            showStatusColumn={false}
```
Comment (tnagorra) [inline src/views/Dashboard/index.tsx:52]:
```suggestion
                            statusColumnShown={false}
```

File: src/config/routes.tsx [hunk 5]
```diff
@@ -177,6 +177,19 @@ const routeSettings = {
             || permissions.actor?.delete
         ),
     }),
+    parkingLots: wrap({
+        path: '/parkinglots/',
```
Comment (tnagorra) [inline src/config/routes.tsx:181]:
```suggestion
        path: '/parking-lot/',
```

File: src/config/routes.tsx [hunk 6]
```diff
@@ -177,6 +177,19 @@ const routeSettings = {
...
+        title: 'Parking Lots',
```
Comment (tnagorra) [inline src/config/routes.tsx:182]:
```suggestion
        title: 'Parking Lot',
```

File: src/config/routes.tsx [hunk 7]
```diff
@@ -177,6 +177,19 @@ const routeSettings = {
...
+        navbarVisibility: true,
+        component: lazy(() => import('../views/ParkingLots')),
```
Comment (tnagorra) [inline src/config/routes.tsx:184]:
```suggestion
        component: lazy(() => import('../views/ParkingLot')),
```

File: src/components/ParkingLotsTable/index.tsx [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+                    className={styles.table}
+                    data={parkingLotsData?.parkingLotList?.results}
+                    keySelector={keySelector}
+                    columns={columns}
+                />
+            )}
+            {(loadingParkingLots || deletingParkingLot) && <Loading absolute />}
+            {!loadingParkingLots && totalParkingLotsCount <= 0 && (
+                <Message
+                    message="No parking lot found."
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:318]:
```suggestion
                    message="No items found."
```

File: src/components/ParkingLotsTable/index.tsx [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+                />
+            )}
+            {shouldShowAddParkingLotModal && (
+                <Modal
+                    onClose={hideAddParkingLotModal}
+                    heading={editableParkingLotId ? 'Edit Parking Lot' : 'Add Parking Lot'}
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:324]:
```suggestion
                    heading={editableParkingLotId ? 'Edit Item' : 'Add Item'}
```

File: src/components/ParkingLotsTable/index.tsx [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+                        placeholder="Search"
+                        onChange={setSearch}
+                    />
+                    {parkingLotPermissions?.add && (
+                        <Button
+                            name={undefined}
+                            onClick={showAddParkingLotModal}
+                            disabled={loadingParkingLots}
+                        >
+                            Add Parking Lot
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:292]:
```suggestion
                            Add Item
```

File: src/components/ParkingLotsTable/index.tsx [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+            parkingLotPermissions?.change,
+        ],
+    );
+    const totalParkingLotsCount = parkingLotsData?.parkingLotList?.totalCount ?? 0;
+
+    return (
+        <Container
+            className={className}
+            contentClassName={styles.content}
+            heading="Parking Lots"
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:275]:
```suggestion
            heading="Parking Lot"
```

File: src/components/ParkingLotsTable/index.tsx [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+            return [
+                createColumn(stringColumn, 'id', 'ID'),
+                createColumn(stringColumn, 'title', 'Title'),
+                showStatusColumn ? createColumn(stringColumn, 'status', 'Status') : undefined,
+                urlColumn,
+                createColumn(stringColumn, 'comments', 'Comments'),
+                actionColumn,
+            ].filter(isDefined);
+        },
+        [
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:260]: Dependency is missing

File: src/components/ParkingLotsTable/index.tsx [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+                cellRendererParams: (_, datum) => ({
+                    id: datum.id,
+                    onDelete: parkingLotPermissions?.delete ? handleParkingLotDelete : undefined,
+                    onEdit: parkingLotPermissions?.change ? showAddParkingLotModal : undefined,
+                }),
+            };
+
+            return [
+                createColumn(stringColumn, 'id', 'ID'),
+                createColumn(stringColumn, 'title', 'Title'),
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:253]:
```suggestion
                createColumn(stringColumn, 'title', 'Title', true),
```

File: src/components/ParkingLotsTable/index.tsx [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+        PARKING_LOT_DELETE,
+        {
+            onCompleted: (response) => {
+                const { deleteParkingLot: deleteParkingLotRes } = response;
+                if (!deleteParkingLotRes) {
+                    return;
+                }
+                const { errors, result } = deleteParkingLotRes;
+                if (errors) {
+                    notify({ children: 'Sorry, Parking lot could not be deleted !' });
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:167]:
```suggestion
                    notify({ children: 'Sorry, Item could not be deleted !' });
```

File: src/components/ParkingLotsTable/index.tsx [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+                }
+                if (result) {
+                    refetchParkingLots(variables);
+                    notify({ children: 'Parking lot deleted successfully!' });
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:171]:
```suggestion
                    notify({ children: 'Item deleted successfully!' });
```

File: src/components/ParkingLotsTable/index.tsx [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+                cellRenderer: ActionCell,
+                cellRendererParams: (_, datum) => ({
+                    id: datum.id,
+                    onDelete: parkingLotPermissions?.delete ? handleParkingLotDelete : undefined,
+                    onEdit: parkingLotPermissions?.change ? showAddParkingLotModal : undefined,
+                }),
+            };
+
+            return [
+                createColumn(stringColumn, 'id', 'ID'),
```
Comment (tnagorra) [inline src/components/ParkingLotsTable/index.tsx:252]: Add createdAt column if default sorting is createdAt

File: src/components/ParkingLotForm/index.tsx [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+    className?: string;
+    onParkingLotCreate?: (result: NonNullable<NonNullable<CreateParkingLotMutation['createParkingLot']>['result']>) => void;
+    id?: string;
+    readOnly?: boolean;
+    onParkingLotFormCancel: () => void;
+}
+
+function ParkingLotForm(props: ParkingLotFormProps) {
+    const {
+        onParkingLotCreate,
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:142]: The problem name can be simpler.

```suggestion
        onCreate,
```

File: src/components/ParkingLotForm/index.tsx [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+        id,
+        readOnly,
+        className,
+        onParkingLotFormCancel,
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:146]:
```suggestion
        onFormCancel,
```

File: src/components/ParkingLotForm/index.tsx [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+                const {
+                    createParkingLot: createParkingLotRes,
+                } = response;
+                if (!createParkingLotRes) {
+                    return;
+                }
+                const { errors, result } = createParkingLotRes;
+                if (errors) {
+                    const formError = transformToFormError(removeNull(errors));
+                    notify({ children: 'Failed to create parkingLot.' });
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:231]:
```suggestion
                    notify({ children: 'Failed to create item.' });
```

File: src/components/ParkingLotForm/index.tsx [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+                    onErrorSet(formError);
+                }
+                if (onParkingLotCreate && result) {
+                    notify({ children: 'ParkingLot created successfully!' });
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:235]:
```suggestion
                    notify({ children: 'Item created successfully!' });
```

File: src/components/ParkingLotForm/index.tsx [hunk 21]
```diff
@@ -0,0 +1,410 @@
...
+                    onPristineSet(true);
+                    onParkingLotCreate(result);
+                }
+            },
+            onError: (errors) => {
+                notify({ children: 'Failed to create parkingLot.' });
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:241]:
```suggestion
                notify({ children: 'Failed to create item.' });
```

File: src/components/ParkingLotForm/index.tsx [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+                const {
+                    updateParkingLot: updateParkingLotRes,
+                } = response;
+                if (!updateParkingLotRes) {
+                    return;
+                }
+                const { errors, result } = updateParkingLotRes;
+                if (errors) {
+                    const formError = transformToFormError(removeNull(errors));
+                    notify({ children: 'Failed to update parkingLot.' });
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:265]:
```suggestion
                    notify({ children: 'Failed to update item.' });
```

File: src/components/ParkingLotForm/index.tsx [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+                    onErrorSet(formError);
+                }
+                if (onParkingLotCreate && result) {
+                    notify({ children: 'ParkingLot updated successfully!' });
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:269]:
```suggestion
                    notify({ children: 'Item updated successfully!' });
```

File: src/components/ParkingLotForm/index.tsx [hunk 24]
```diff
@@ -0,0 +1,410 @@
...
+                    onPristineSet(true);
+                    onParkingLotCreate(result);
+                }
+            },
+            onError: (errors) => {
+                notify({ children: 'Failed to update parkingLot.' });
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:275]:
```suggestion
                notify({ children: 'Failed to update item.' });
```

File: src/components/ParkingLotForm/index.tsx [hunk 25]
```diff
@@ -0,0 +1,410 @@
...
+                    onOptionsChange={setCountryOptions}
+                    onChange={onValueChange}
+                    value={value.country}
+                    error={error?.fields?.country}
+                    disabled={disabled}
+                />
+            </div>
+            <div className={styles.row}>
+                <UserSelectInput
+                    label="Assign To *"
```
Comment (tnagorra) [inline src/components/ParkingLotForm/index.tsx:349]:
```suggestion
                    label="Assignee *"
```

**Review by tnagorra — CHANGES_REQUESTED** [review]

**Review by frozenhelium — APPROVED** [review]

**Pull request merged**